File: workbench/min0_core_forth_linker.rb
# frozen_string_literal: true

module Min0CoreForth
  MANIFEST_FORMAT = "min0-core-forth-relocation-manifest"
  MANIFEST_VERSION = 1
  MANIFEST_PROFILE = "reference32-le"
  LINK_SECTIONS = ["code", "dictionary", "data"].freeze
  REFERENCE32_LIMIT = 1 << 32

  class LinkError < StandardError; end

  module Linker
    module_function

    def build_manifest(records)
      {
        format: MANIFEST_FORMAT,
        version: MANIFEST_VERSION,
        profile: MANIFEST_PROFILE,
        records: records.map(&:dup)
      }
    end

    def link_components(components, source_bases, target_bases, manifest)
      images = normalize_components(components)
      old_bases = normalize_bases(source_bases, "source")
      new_bases = normalize_bases(target_bases, "target")
      validate_ranges(images, old_bases, "source")
      validate_ranges(images, new_bases, "target")
      raise LinkError, "manifest must be a Hash" unless manifest.is_a?(Hash)
      unless manifest.keys.map(&:to_s).sort == %w[format profile records version]
        raise LinkError, "relocation manifest fields are malformed"
      end
      unless field(manifest, :format) == MANIFEST_FORMAT
        raise LinkError, "unsupported relocation manifest format"
      end
      unless integer(field(manifest, :version), "manifest version") == MANIFEST_VERSION
        raise LinkError, "unsupported relocation manifest version"
      end
      unless field(manifest, :profile) == MANIFEST_PROFILE
        raise LinkError, "unsupported relocation manifest profile"
      end
      records = field(manifest, :records)
      raise LinkError, "manifest records must be an Array" unless records.is_a?(Array)

      occupied = LINK_SECTIONS.to_h { |section| [section, []] }
      patches = []
      records.each_with_index do |record, index|
        prefix = "relocation record #{index}"
        raise LinkError, "#{prefix} must be a Hash" unless record.is_a?(Hash)
        unless record.keys.map(&:to_s).sort == %w[kind offset section target width]
          raise LinkError, "#{prefix} fields are malformed"
        end
        section = field(record, :section)
        target = field(record, :target)
        raise LinkError, "#{prefix} has unknown patch section" unless LINK_SECTIONS.include?(section)
        raise LinkError, "#{prefix} has unknown target section" unless LINK_SECTIONS.include?(target)

        offset = integer(field(record, :offset), "#{prefix} offset")
        width = integer(field(record, :width), "#{prefix} width")
        kind = field(record, :kind)
        unless kind.is_a?(String) && !kind.empty?
          raise LinkError, "#{prefix} kind must be a nonempty String"
        end
        unless width == 4
          raise LinkError, "#{prefix} width is unsupported for Reference32"
        end
        if offset.negative? || offset + width > images.fetch(section).bytesize
          raise LinkError, "#{prefix} patch is outside its component"
        end
        if section == "dictionary" && ((old_bases.fetch(section) + offset) % 4).nonzero?
          raise LinkError, "#{prefix} dictionary patch is not cell-aligned"
        end
        interval = [offset, offset + width]
        if occupied.fetch(section).any? { |start, finish| interval[0] < finish && start < interval[1] }
          raise LinkError, "#{prefix} overlaps another patch"
        end
        occupied.fetch(section) << interval

        old_value = images.fetch(section).byteslice(offset, width).unpack1("V")
        target_start = old_bases.fetch(target)
        target_end = target_start + images.fetch(target).bytesize
        in_source = if target == "data"
                      old_value.between?(target_start, target_end)
                    else
                      old_value >= target_start && old_value < target_end
                    end
        raise LinkError, "#{prefix} value is outside its source target component" unless in_source

        new_value = old_value + new_bases.fetch(target) - old_bases.fetch(target)
        unless new_value >= 0 && new_value < REFERENCE32_LIMIT
          raise LinkError, "#{prefix} result is outside Reference32"
        end
        moved_end = new_bases.fetch(target) + images.fetch(target).bytesize
        in_target = if target == "data"
                      new_value.between?(new_bases.fetch(target), moved_end)
                    else
                      new_value >= new_bases.fetch(target) && new_value < moved_end
                    end
        raise LinkError, "#{prefix} result is outside its target component" unless in_target

        patches << [section, offset, new_value]
      end

      result = images.transform_values(&:dup)
      patches.each do |section, offset, value|
        result.fetch(section)[offset, 4] = [value].pack("V")
      end
      result
    end

    def field(hash, name)
      hash.key?(name) ? hash[name] : hash[name.to_s]
    end
    private_class_method :field

    def integer(value, label)
      raise LinkError, "#{label} must be an Integer" unless value.is_a?(Integer)

      value
    end
    private_class_method :integer

    def normalize_components(components)
      raise LinkError, "components must be a Hash" unless components.is_a?(Hash)

      LINK_SECTIONS.to_h do |section|
        value = field(components, section.to_sym)
        raise LinkError, "component #{section} must be a String" unless value.is_a?(String)

        [section, value.b]
      end
    end
    private_class_method :normalize_components

    def normalize_bases(bases, label)
      raise LinkError, "#{label} bases must be a Hash" unless bases.is_a?(Hash)

      LINK_SECTIONS.to_h do |section|
        base = integer(field(bases, section.to_sym), "#{label} base #{section}")
        unless base >= 0 && base < REFERENCE32_LIMIT
          raise LinkError, "#{label} base #{section} is outside Reference32"
        end
        [section, base]
      end
    end
    private_class_method :normalize_bases

    def validate_ranges(images, bases, label)
      ranges = LINK_SECTIONS.filter_map do |section|
        start = bases.fetch(section)
        finish = start + images.fetch(section).bytesize
        raise LinkError, "#{label} component #{section} exceeds Reference32" if finish > REFERENCE32_LIMIT

        [start, finish, section] if finish > start
      end.sort
      ranges.each_cons(2) do |previous, current|
        if current[0] < previous[1]
          raise LinkError, "#{label} components #{previous[2]} and #{current[2]} overlap"
        end
      end
    end
    private_class_method :validate_ranges
  end
end

File: workbench/min0_core_forth_linker.py
"""Validated copy-on-success relocation linker for MIN0 CORE FORTH R0 images."""

from __future__ import annotations

from collections.abc import Mapping


MANIFEST_FORMAT = "min0-core-forth-relocation-manifest"
MANIFEST_VERSION = 1
MANIFEST_PROFILE = "reference32-le"
SECTIONS = ("code", "dictionary", "data")
REFERENCE32_LIMIT = 1 << 32


class LinkError(ValueError):
    pass


def build_manifest(records: list[dict[str, str | int]]) -> dict:
    """Build the current in-memory manifest envelope without freezing a file format."""

    return {
        "format": MANIFEST_FORMAT,
        "version": MANIFEST_VERSION,
        "profile": MANIFEST_PROFILE,
        "records": [dict(record) for record in records],
    }


def _integer(value: object, label: str) -> int:
    if not isinstance(value, int) or isinstance(value, bool):
        raise LinkError(f"{label} must be an integer")
    return value


def _normalize_components(components: Mapping[str, object]) -> dict[str, bytes]:
    if not isinstance(components, Mapping):
        raise LinkError("components must be a mapping")
    result: dict[str, bytes] = {}
    for section in SECTIONS:
        value = components.get(section)
        if not isinstance(value, (bytes, bytearray, memoryview)):
            raise LinkError(f"component {section} must be bytes")
        result[section] = bytes(value)
    return result


def _normalize_bases(bases: Mapping[str, object], label: str) -> dict[str, int]:
    if not isinstance(bases, Mapping):
        raise LinkError(f"{label} bases must be a mapping")
    result: dict[str, int] = {}
    for section in SECTIONS:
        base = _integer(bases.get(section), f"{label} base {section}")
        if base < 0 or base >= REFERENCE32_LIMIT:
            raise LinkError(f"{label} base {section} is outside Reference32")
        result[section] = base
    return result


def _validate_ranges(
    components: dict[str, bytes], bases: dict[str, int], label: str
) -> None:
    ranges: list[tuple[int, int, str]] = []
    for section in SECTIONS:
        start = bases[section]
        end = start + len(components[section])
        if end > REFERENCE32_LIMIT:
            raise LinkError(f"{label} component {section} exceeds Reference32")
        if end > start:
            ranges.append((start, end, section))
    ranges.sort()
    for previous, current in zip(ranges, ranges[1:]):
        if current[0] < previous[1]:
            raise LinkError(
                f"{label} components {previous[2]} and {current[2]} overlap"
            )


def link_components(
    components: Mapping[str, object],
    source_bases: Mapping[str, object],
    target_bases: Mapping[str, object],
    manifest: object,
) -> dict[str, bytes]:
    """Validate every record, then patch private copies and return them."""

    images = _normalize_components(components)
    old_bases = _normalize_bases(source_bases, "source")
    new_bases = _normalize_bases(target_bases, "target")
    _validate_ranges(images, old_bases, "source")
    _validate_ranges(images, new_bases, "target")
    if not isinstance(manifest, Mapping):
        raise LinkError("manifest must be a mapping")
    if set(manifest) != {"format", "version", "profile", "records"}:
        raise LinkError("relocation manifest fields are malformed")
    if manifest.get("format") != MANIFEST_FORMAT:
        raise LinkError("unsupported relocation manifest format")
    if _integer(manifest.get("version"), "manifest version") != MANIFEST_VERSION:
        raise LinkError("unsupported relocation manifest version")
    if manifest.get("profile") != MANIFEST_PROFILE:
        raise LinkError("unsupported relocation manifest profile")
    records = manifest.get("records")
    if not isinstance(records, list):
        raise LinkError("manifest records must be a list")

    occupied: dict[str, list[tuple[int, int]]] = {section: [] for section in SECTIONS}
    patches: list[tuple[str, int, int]] = []
    for index, record in enumerate(records):
        prefix = f"relocation record {index}"
        if not isinstance(record, Mapping):
            raise LinkError(f"{prefix} must be a mapping")
        if set(record) != {"section", "offset", "target", "width", "kind"}:
            raise LinkError(f"{prefix} fields are malformed")
        section = record.get("section")
        target = record.get("target")
        if section not in SECTIONS:
            raise LinkError(f"{prefix} has unknown patch section")
        if target not in SECTIONS:
            raise LinkError(f"{prefix} has unknown target section")
        offset = _integer(record.get("offset"), f"{prefix} offset")
        width = _integer(record.get("width"), f"{prefix} width")
        kind = record.get("kind")
        if not isinstance(kind, str) or not kind:
            raise LinkError(f"{prefix} kind must be a nonempty string")
        if width != 4:
            raise LinkError(f"{prefix} width is unsupported for Reference32")
        if offset < 0 or offset + width > len(images[section]):
            raise LinkError(f"{prefix} patch is outside its component")
        if section == "dictionary" and (old_bases[section] + offset) % 4:
            raise LinkError(f"{prefix} dictionary patch is not cell-aligned")
        interval = (offset, offset + width)
        if any(interval[0] < end and start < interval[1] for start, end in occupied[section]):
            raise LinkError(f"{prefix} overlaps another patch")
        occupied[section].append(interval)

        old_value = int.from_bytes(
            images[section][offset : offset + width], "little", signed=False
        )
        target_start = old_bases[target]
        target_end = target_start + len(images[target])
        in_source = (
            target_start <= old_value <= target_end
            if target == "data"
            else target_start <= old_value < target_end
        )
        if not in_source:
            raise LinkError(f"{prefix} value is outside its source target component")
        new_value = old_value + new_bases[target] - old_bases[target]
        if new_value < 0 or new_value >= REFERENCE32_LIMIT:
            raise LinkError(f"{prefix} result is outside Reference32")
        moved_end = new_bases[target] + len(images[target])
        in_target = (
            new_bases[target] <= new_value <= moved_end
            if target == "data"
            else new_bases[target] <= new_value < moved_end
        )
        if not in_target:
            raise LinkError(f"{prefix} result is outside its target component")
        patches.append((section, offset, new_value))

    result = {section: bytearray(image) for section, image in images.items()}
    for section, offset, value in patches:
        result[section][offset : offset + 4] = value.to_bytes(4, "little")
    return {section: bytes(image) for section, image in result.items()}
